monkey/agent_plugins/exploiters/ssh/src/ssh_client.py

```diff
@@ -0,0 +1,155 @@
+            )
+        except paramiko.AuthenticationError as err:
+            self._client.close()
+            error_message = (
+                f"Failed logging into victim {host.ip} with {username}@{host.ip}"
+                f"private key: {err}"
+            )
+            logger.info(error_message)
+            raise err
+        except Exception as err:
+            error_message = (
+                f"Unexpected error while attempting to login to {username}@{host.ip} with ssh key: "
+                f"{err}"
+            )
+            logger.error(error_message)
+            self._client.close()
+            raise err
+
+    def _connect_with_login_credentials(
+        self, host: TargetHost, credentials: Credentials, port: NetworkPort, timeout: float
+    ):
+        username = (
+            credentials.identity.username
+            if isinstance(credentials.identity, Username)
+            else credentials.identity
+        )
+
+        try:
+            self._client.connect(
+                str(host.ip),
+                username=username,
+                password=get_plaintext(credentials.secret.password),
+                port=int(port),
+                timeout=timeout,
+                auth_timeout=SSH_AUTH_TIMEOUT,
+                banner_timeout=SSH_BANNER_TIMEOUT,
+                channel_timeout=SSH_CHANNEL_TIMEOUT,
+                allow_agent=False,
+            )
+            logger.debug(f"Successfully logged in {host.ip}, User: {username}")
+        except paramiko.AuthenticationException as err:
+            self._client.close()
+            error_message = f"Failed logging into victim {host.ip} with user: {username}: {err}"
+            raise err
+        except Exception as err:
+            error_message = (
+                f"Unexpected error while attempting to login to {host.ip} with password: " f"{err}"
+            )
+            logger.debug(error_message)
+            raise err
+
+    def copy_file(
+        self,
+        binary_file_object: io.BytesIO,
+        destination_path: PurePath,
+    ):
+        try:
+            with self._client.open_sftp() as sftp:
+                sftp.putfo(
+                    binary_file_object,
+                    str(destination_path),
+                    file_size=len(binary_file_object.getbuffer()),
+                    callback=self._log_transfer,
+                )
+                sftp.chmod(str(destination_path), 0o700)
+        except Exception as err:
+            error_message = f"Error uploading file: ({err})"
+            logger.error(error_message)
+            raise err
+
+    def _log_transfer(self, transferred: int, total: int):
+        timer = EggTimer()
+        timer.set(TRANSFER_UPDATE_RATE)
+
+        if timer.is_expired():
+            logger.debug(f"SFTP transferred: {transferred} bytes, total: {total} bytes")
+            timer.reset()
+
+    def execute_command(self, command: str) -> bytes:
```

Comment:
Can we document the public interface for this class? This method can raise an error, but it's not documented.

Reply:
This will be addressed in the next PR